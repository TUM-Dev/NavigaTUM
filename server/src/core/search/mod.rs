mod search_executor;

use actix_web::{get, web, HttpRequest, HttpResponse};
use serde::{Deserialize, Serialize};
use std::time::Instant;
#[derive(Deserialize)]
pub struct SearchQueryArgs {
    q: String,
    // Limit per facet
    limit_buildings: Option<usize>,
    limit_rooms: Option<usize>,
    limit_all: Option<usize>,
}

/// Returned search results by this
#[derive(Serialize, Debug)]
pub struct SearchResults {
    sections: Vec<search_executor::SearchResultsSection>,
    time_ms: u128,
}

#[derive(Deserialize, Clone, Eq, PartialEq, Hash)]
pub struct SanitisedSearchQueryArgs {
    // Limit per facet
    pub limit_buildings: usize,
    pub limit_rooms: usize,
    pub limit_all: usize,
}

#[get("/search")]
pub async fn search_handler(
    _req: HttpRequest,
    web::Query(args): web::Query<SearchQueryArgs>,
) -> HttpResponse {
    let start_time = Instant::now();

    let (q, sanitised_args) = sanitise_args(args);
    let results_sections = search_executor::do_geoentry_search(q, sanitised_args).await;
    let time_ms = start_time.elapsed().as_millis();

    if results_sections.len() != 2 {
        return HttpResponse::InternalServerError().body("Internal error");
    }
    let search_results = SearchResults {
        sections: results_sections,
        time_ms,
    };
    return HttpResponse::Ok().json(search_results);
}

fn sanitise_args(args: SearchQueryArgs) -> (String, SanitisedSearchQueryArgs) {
    let sanitised_args = SanitisedSearchQueryArgs {
        limit_buildings: args.limit_buildings.unwrap_or(5).clamp(0, 1_000),
        limit_rooms: args.limit_rooms.unwrap_or(10).clamp(0, 1_000),
        limit_all: args.limit_all.unwrap_or(10).clamp(1, 1_000),
    };
    return (args.q, sanitised_args);
}
