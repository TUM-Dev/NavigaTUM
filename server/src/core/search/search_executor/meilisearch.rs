use awc::Client;

use crate::core::search::search_executor::preprocess::SearchToken;
use serde::{Deserialize, Serialize};
use serde_json::Result;
use std::collections::HashMap;

/// Input into MeiliSearch
pub(super) struct MSSearchArgs<'a> {
    pub(super) q: &'a str,
    pub(super) filter: Option<MSSearchFilter>,
    pub(super) limit: usize,
}

#[derive(Debug)]
pub(super) struct MSSearchFilter {
    facet: Vec<String>,
}

#[derive(Serialize)]
struct MSQuery<'a> {
    q: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    filter: Option<Vec<Vec<String>>>,
    limit: usize,
    #[serde(rename = "facetsDistribution")]
    facets_distribution: Vec<String>,
}

/// Result format of MeiliSearch.
#[derive(Deserialize)]
#[allow(dead_code)]
pub(super) struct MSResults {
    pub(super) hits: Vec<MSHit>,
    #[serde(rename = "nbHits")]
    pub(super) nb_hits: i32,
    #[serde(rename = "facetsDistribution")]
    pub(super) facets_distribution: MSFacetDistribution,
}

#[derive(Deserialize, Debug)]
pub(super) struct MSFacetDistribution {
    pub(super) facet: HashMap<String, i32>,
}

#[derive(Deserialize, Clone)]
#[allow(dead_code)]
pub(super) struct MSHit {
    ms_id: String,
    pub(super) id: String,
    pub(super) name: String,
    pub(super) arch_name: Option<String>,
    pub(super) r#type: String,
    pub(super) type_common_name: String,
    pub(super) parent_building: Vec<String>,
    parent_keywords: Vec<String>,
    address: Option<String>,
    usage: Option<String>,
    rank: i32,
}

pub(super) async fn do_meilisearch(client: Client, args: MSSearchArgs<'_>) -> Result<MSResults> {
    let post_data = MSQuery {
        q: args.q,
        filter: match args.filter {
            Some(f) => {
                let mut f_array = Vec::<Vec<String>>::new();
                // Currently only facets, but later also more filters
                let mut sub_f_array = Vec::<String>::new();
                for facet in f.facet {
                    sub_f_array.push(format!("facet = {}", facet)); // TODO: Put in quotes?
                }
                f_array.push(sub_f_array);

                Some(f_array)
            }
            _ => None,
        },
        limit: args.limit,
        facets_distribution: vec!["facet".to_string()],
    };

    let resp_bytes = client
        .post(
            std::env::var("MEILISEARCH_URL")
                .unwrap_or_else(|_| "http://localhost:7700/indexes/entries/search".to_string()),
        )
        .send_json(&post_data)
        .await
        .unwrap()
        .body()
        .await
        .unwrap();

    let resp_str = std::str::from_utf8(resp_bytes.as_ref()).unwrap();
    Ok(serde_json::from_str(resp_str)?)
}

pub(super) async fn do_building_search_closed_query(
    client: Client,
    query_string: String,
    limit: usize,
) -> Result<MSResults> {
    do_meilisearch(
        client,
        MSSearchArgs {
            q: &query_string,
            filter: Some(MSSearchFilter {
                facet: vec!["site".to_string(), "building".to_string()],
            }),
            limit,
        },
    )
    .await
}

pub(super) async fn do_room_search(
    client: Client,
    search_tokens: &Vec<SearchToken>,
    limit: usize,
) -> Result<MSResults> {
    let mut q = String::from("");
    for token in search_tokens {
        // It is common that room names are given with four-digits with the first digit
        // being the level. In this case we add splitted terms search well, which could give
        // results if the 4-digit-token doesn't, but still the 4-digit-token should usually
        // take precedence.
        let s = if token.s.len() == 4
            && match token.s.chars().next().unwrap() {
                '0' | '1' | '2' => true,
                _ => false,
            }
            && token.s.chars().all(char::is_numeric)
        {
            format!(
                "{} {} {}",
                token.s,
                token.s.get(0..1).unwrap(),
                token.s.get(1..4).unwrap()
            )
        } else {
            token.s.clone()
        };

        if token.closed && !token.quoted {
            q.push_str(&format!("{} ", s));
        } else {
            q.push_str(&s);
        }
    }

    do_meilisearch(
        client,
        MSSearchArgs {
            q: &q,
            filter: Some(MSSearchFilter {
                facet: vec!["room".to_string()],
            }),
            limit,
        },
    )
    .await
}
