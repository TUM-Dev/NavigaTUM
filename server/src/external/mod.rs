pub mod connectum;
pub mod download_map_image;
pub mod github;
pub mod meilisearch;
pub mod nominatim;
pub mod valhalla;
