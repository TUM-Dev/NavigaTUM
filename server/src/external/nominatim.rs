use crate::limited::vec::LimitedVec;
use serde::Deserialize;

#[derive(Deserialize, Clone)]
pub struct NominatimAddressResponse {
    // postcode: Option<String>,
    // country: Option<String>,
    // country_code: Option<String>,
    // ISO3166-2-lvl4: Option<String>,
    state: Option<String>,
    county: Option<String>,
    town: Option<String>,
    suburb: Option<String>,
    village: Option<String>,
    hamlet: Option<String>,
    pub road: Option<String>,
}

impl NominatimAddressResponse {
    pub fn serialise(&self) -> String {
        let mut result = Vec::<String>::new();
        if let Some(state) = self.state.clone() {
            result.push(state);
        }
        if let Some(county) = self.county.clone() {
            result.push(county);
        }
        if let Some(town) = self.town.clone() {
            result.push(town);
        }
        if let Some(suburb) = self.suburb.clone() {
            result.push(suburb);
        }
        if let Some(village) = self.village.clone() {
            result.push(village);
        }
        if let Some(hamlet) = self.hamlet.clone() {
            result.push(hamlet);
        }
        if let Some(road) = self.road.clone() {
            result.push(road);
        }
        result.join(", ")
    }
}

#[derive(Deserialize, Clone)]
pub struct Nominatim {
    /// Example: 371651568
    pub osm_id: i64,
    /// Example: "road",
    #[serde(rename = "addresstype")]
    pub address_type: String,
    /// Example: "Münchner Straße",
    pub name: String,
    pub address: NominatimAddressResponse,
}
impl Nominatim {
    #[tracing::instrument]
    pub async fn address_search(q: &str) -> anyhow::Result<LimitedVec<Self>> {
        let url = std::env::var("NOMINATIM_URL")
            .unwrap_or_else(|_| "https://nav.tum.de/nominatim".to_string());
        let url = format!("{url}/search?q={q}&addressdetails=1");
        let Ok(nominatim_results) = reqwest::get(&url).await else {
            anyhow::bail!("cannot get {url}");
        };
        let Ok(results) = nominatim_results.json::<Vec<Self>>().await else {
            anyhow::bail!("the results from nomnatim is not what we expected {url}");
        };
        Ok(LimitedVec(results))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn serialize_address() {
        let response = NominatimAddressResponse {
            state: None,
            county: None,
            town: None,
            suburb: None,
            village: None,
            hamlet: None,
            road: None,
        };
        insta::assert_snapshot!(response.serialise(), @"");
        let response = NominatimAddressResponse {
            state: Some("Bavaria".to_string()),
            county: Some("Germany".to_string()),
            town: Some("Berlin".to_string()),
            suburb: Some("Neuköln".to_string()),
            village: None,
            hamlet: None,
            road: Some("Münchnerstraße 21".to_string()),
        };
        insta::assert_snapshot!(response.serialise(), @"Bavaria, Germany, Berlin, Neuköln, Münchnerstraße 21");
    }
}
