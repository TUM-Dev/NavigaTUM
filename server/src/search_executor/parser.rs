use logos::Logos;
use std::collections::HashSet;
use std::fmt::{Debug, Formatter};
use tracing::warn;

use super::lexer::Token;

#[derive(Clone, Default, PartialEq, Eq)]
pub struct Filter {
    parents: HashSet<String>,
    types: HashSet<String>,
    usages: HashSet<String>,
}
impl Filter {
    pub fn as_meilisearch_filters(&self) -> String {
        let mut filters = vec![];
        if !self.parents.is_empty() {
            let parents: Vec<&str> = self.parents.iter().map(String::as_str).collect();
            filters.push(format!(
                "((parent_keywords IN {parents:?}) OR (parent_building_names IN {parents:?}) OR (campus IN {parents:?}))"
            ));
        }
        if !self.types.is_empty() {
            let types: Vec<&str> = self.types.iter().map(String::as_str).collect();
            filters.push(format!("(type IN {types:?})"));
        }
        if !self.usages.is_empty() {
            let usages: Vec<&str> = self.usages.iter().map(String::as_str).collect();
            filters.push(format!("(usage IN {usages:?})"));
        }
        filters.join(" AND ")
    }
    pub fn is_empty(&self) -> bool {
        self.parents.is_empty() && self.types.is_empty() && self.usages.is_empty()
    }
}

impl Debug for Filter {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("Filter");
        if !self.parents.is_empty() {
            base.field("parents", &self.parents);
        }
        if !self.types.is_empty() {
            base.field("types", &self.parents);
        }
        if !self.usages.is_empty() {
            base.field("usages", &self.parents);
        }
        base.finish()
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Sorting {
    location: HashSet<String>,
}

impl Sorting {
    pub fn as_meilisearch_sorting(&self) -> Vec<String> {
        self.location
            .iter()
            .map(|s| format!("_geoPoint({s}):asc"))
            .collect()
    }
    pub fn is_empty(&self) -> bool {
        self.location.is_empty()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TextToken {
    Text(String),
    SplittableText((String, String)),
}

#[derive(Clone, Default, PartialEq, Eq)]
pub struct ParsedQuery {
    pub tokens: Vec<TextToken>,
    pub filters: Filter,
    pub sorting: Sorting,
}

impl Debug for ParsedQuery {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("ParsedQuery");
        base.field("tokens", &self.tokens);
        if !self.filters.is_empty() {
            base.field("from", &self.filters);
        }
        if !self.sorting.is_empty() {
            base.field("from", &self.sorting);
        }
        base.finish()
    }
}

impl ParsedQuery {
    pub fn relevant_enough_for_room_highligting(&self) -> bool {
        if self.tokens.len() == 1 {
            return true;
        }
        match self.tokens.first() {
            Some(first) => match first {
                TextToken::Text(t) => t.len() > 3,
                TextToken::SplittableText((t0, _)) => t0.len() > 3,
            },
            None => false,
        }
    }
}

impl From<&str> for ParsedQuery {
    fn from(query: &str) -> Self {
        let mut result = Self::default();
        for token in Token::lexer(query) {
            match token {
                Ok(Token::Text(s)) => {
                    result.tokens.push(TextToken::Text(s));
                }
                Ok(Token::SplittableText((s1, s2))) => {
                    result.tokens.push(TextToken::SplittableText((s1, s2)));
                }
                Ok(Token::ParentFilter(filter)) => {
                    result.filters.parents.insert(filter);
                }
                Ok(Token::UsageFilter(filter)) => {
                    result.filters.usages.insert(filter);
                }
                Ok(Token::TypeFilter(filter)) => {
                    result.filters.types.insert(filter);
                }
                Ok(Token::LocationSort(location)) => {
                    result.sorting.location.insert(location);
                }
                Err(_) => {
                    warn!("Error in query parsing");
                }
            }
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn parent_filter() {
        for filter in ["in:", "@"] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        parents: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn usage_filters() {
        for filter in ["usage:", "nutzung:", "="] {
            for sep in ["", " "] {
                assert_eq!(
                    ParsedQuery::from(format!("{filter}{sep}foo").as_str()).filters,
                    Filter {
                        usages: HashSet::from(["foo".to_string()]),
                        ..Default::default()
                    }
                );
            }
        }
    }

    #[test]
    fn type_filters() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("type:{sep}foo").as_str()).filters,
                Filter {
                    types: HashSet::from(["foo".to_string()]),
                    ..Default::default()
                }
            );
        }
    }

    #[test]
    fn location_sort() {
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.32,59.3").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.32,59.3".to_string()]),
                }
            );
        }
        for sep in ["", " "] {
            assert_eq!(
                ParsedQuery::from(format!("near:{sep}45.3,59.00000003").as_str()).sorting,
                Sorting {
                    location: HashSet::from(["45.3,59.00000003".to_string()]),
                }
            );
        }
    }

    #[test]
    fn text_token() {
        assert_eq!(
            ParsedQuery::from("foo").tokens,
            vec![TextToken::Text("foo".to_string())]
        );
        assert_eq!(
            ParsedQuery::from("foo foo").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from("foo hs1").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::SplittableText(("hs".to_string(), "1".to_string())),
            ]
        );
    }

    #[test]
    fn text_filter_mixed() {
        assert_eq!(
            ParsedQuery::from("foo in:abc bar @abc foo near:45.32,59.3").tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
        assert_eq!(
            ParsedQuery::from(
                "foo in:abc bar @abc =def usage:dd nutzung:gh type:fdh foo near:45.32,59.3"
            )
            .tokens,
            vec![
                TextToken::Text("foo".to_string()),
                TextToken::Text("bar".to_string()),
                TextToken::Text("foo".to_string()),
            ]
        );
    }
}
