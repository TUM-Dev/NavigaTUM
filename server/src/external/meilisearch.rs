use meilisearch_sdk::client::Client;
use meilisearch_sdk::errors::Error;
use meilisearch_sdk::indexes::Index;
use meilisearch_sdk::search::{MultiSearchResponse, SearchQuery, Selectors};
use serde::Deserialize;
use std::fmt::{Debug, Formatter};

use crate::search::{Highlighting, Limits};

#[derive(Deserialize, Default, Clone)]
#[allow(dead_code)]
pub struct MSHit {
    ms_id: String,
    pub id: String,
    pub name: String,
    pub arch_name: Option<String>,
    pub r#type: String,
    pub type_common_name: String,
    pub parent_building_names: Vec<String>,
    parent_keywords: Vec<String>,
    pub campus: Option<String>,
    address: Option<String>,
    usage: Option<String>,
    rank: i32,
}
impl Debug for MSHit {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("MSHit")
            .field("id", &self.id)
            .field("name", &self.name)
            .finish()
    }
}

#[derive(Clone)]
struct GeoEntryFilters {
    default: String,
    rooms: String,
    buildings: String,
}
impl Default for GeoEntryFilters {
    fn default() -> Self {
        Self {
            default: "".to_string(),
            rooms: "facet = \"room\"".to_string(),
            buildings: "facet = \"building\"".to_string(),
        }
    }
}
impl Debug for GeoEntryFilters {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("GeoEntryFilters");
        if !self.default.is_empty() {
            base.field("default", &self.default);
        }
        if !self.rooms.is_empty() {
            base.field("rooms", &self.rooms);
        }
        if !self.buildings.is_empty() {
            base.field("buildings", &self.buildings);
        }
        base.finish()
    }
}

impl From<&String> for GeoEntryFilters {
    fn from(ms_filter: &String) -> Self {
        Self::default().with_filter(ms_filter)
    }
}

impl GeoEntryFilters {
    pub fn with_filter(&mut self, ms_filter: impl ToString) -> Self {
        let ms_filter = ms_filter.to_string();
        // --
        if self.default.is_empty() && !ms_filter.is_empty() {
            self.default.push_str(" AND ")
        }
        self.default.push_str(&ms_filter);
        // --
        if self.buildings.is_empty() && !ms_filter.is_empty() {
            self.buildings.push_str(" AND ")
        }
        self.buildings.push_str(&ms_filter);
        // --
        if self.rooms.is_empty() && !ms_filter.is_empty() {
            self.rooms.push_str(" AND ")
        }
        self.rooms.push_str(&ms_filter);
        self.clone()
    }
}

#[derive(Clone)]
pub struct GeoEntryQuery {
    client: Client,
    query: String,
    limits: Limits,
    highlighting: Highlighting,
    filters: GeoEntryFilters,
    sorting: Vec<String>,
}

impl From<(&Client, String, &Limits, &Highlighting)> for GeoEntryQuery {
    fn from(
        (client, query, limits, highlighting): (&Client, String, &Limits, &Highlighting),
    ) -> Self {
        Self {
            client: client.clone(),
            query,
            limits: *limits,
            highlighting: highlighting.clone(),
            filters: GeoEntryFilters::default(),
            sorting: Vec::new(),
        }
    }
}

impl GeoEntryQuery {
    // add sorting constraints
    pub fn with_sorting(&mut self, sortation: impl ToString) -> Self {
        self.sorting.push(sortation.to_string());
        self.clone()
    }
    // add filtering constraints
    pub fn with_filtering(&mut self, ms_filter: impl ToString) -> Self {
        self.filters.with_filter(ms_filter);
        self.clone()
    }
    pub async fn execute(self) -> Result<MultiSearchResponse<MSHit>, Error> {
        let entries = self.client.index("entries");

        // due to lifetime shenanigans this is added here (I can't make it move down to the other statements)
        // If you can make it, please propose a PR, I know that this is really hacky ^^
        let sorting = self
            .sorting
            .iter()
            .map(String::as_str)
            .collect::<Vec<&str>>();

        // Currently ranking is designed to put buildings at the top if they equally
        // match the term compared to a room. For this reason there is only a search
        // for all entries and only rooms, search matching (and relevant) buildings can be
        // expected to be at the top of the merged search. However sometimes a lot of
        // buildings will be hidden (e.g. building parts), so the extra room search ....
        self.client
            .multi_search()
            .with_search_query(
                self.merged_query(&entries, &self.query)
                    .with_sort(&sorting)
                    .build(),
            )
            .with_search_query(
                self.buildings_query(&entries, &self.query)
                    .with_sort(&sorting)
                    .build(),
            )
            .with_search_query(
                self.rooms_query(&entries, &self.query)
                    .with_sort(&sorting)
                    .build(),
            )
            .execute::<MSHit>()
            .await
    }

    fn common_query<'b: 'a, 'a>(
        &'b self,
        entries: &'a Index,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        SearchQuery::new(entries)
            .with_facets(Selectors::Some(&["facet"]))
            .with_highlight_pre_tag(&self.highlighting.pre)
            .with_highlight_post_tag(&self.highlighting.post)
            .with_attributes_to_highlight(Selectors::Some(&["name"]))
            .build()
    }

    fn merged_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        let mut s = self
            .common_query(entries)
            .with_query(query)
            .with_limit(self.limits.total_count)
            .build();
        if !self.filters.default.is_empty() {
            s = s.with_filter(&self.filters.default).build();
        }
        s
    }

    fn buildings_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        self.common_query(entries)
            .with_query(query)
            .with_limit(2 * self.limits.buildings_count) // we might do reordering later
            .with_filter(&self.filters.buildings)
            .build()
    }

    fn rooms_query<'a>(
        &'a self,
        entries: &'a Index,
        query: &'a str,
    ) -> SearchQuery<'a, meilisearch_sdk::DefaultHttpClient> {
        self.common_query(entries)
            .with_query(query)
            .with_limit(self.limits.rooms_count)
            .with_filter(&self.filters.rooms)
            .build()
    }
}

impl Debug for GeoEntryQuery {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_struct("GeoEntryQuery");
        base.field("query", &self.query)
            .field("limits", &self.limits)
            .field("highlighting", &self.highlighting)
            .field("filters", &self.filters);
        if !self.sorting.is_empty() {
            base.field("sorting", &self.sorting);
        }
        base.finish()
    }
}
