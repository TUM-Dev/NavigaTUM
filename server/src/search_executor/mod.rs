use meilisearch_sdk::client::Client;
use parser::TextToken;
use serde::Serialize;
use std::fmt::{Debug, Formatter};
use tracing::error;

use crate::external::meilisearch::{GeoEntryQuery, MSHit};
use crate::external::nominatim::Nominatim;
use crate::limited::vec::LimitedVec;
use crate::routes::search::{Highlighting, Limits};
use crate::search_executor::parser::ParsedQuery;

mod formatter;
mod lexer;
mod merger;
mod parser;

#[derive(Serialize, Clone, Copy, utoipa::ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum ResultFacet {
    SitesBuildings,
    Rooms,
    Addresses,
}

#[derive(Serialize, Clone, utoipa::ToSchema)]
pub struct ResultsSection {
    /// These indicate the type of item this represents
    pub(crate) facet: ResultFacet,
    entries: Vec<ResultEntry>,
    /// A recommendation how many of the entries should be displayed by default.
    ///
    /// The number is usually from `0`..`5`.
    /// More results might be displayed when clicking "expand".
    #[schema(example = 4)]
    n_visible: usize,
    /// The estimated (not exact) number of hits for that query
    #[serde(rename = "estimatedTotalHits")]
    #[schema(example = 6)]
    estimated_total_hits: usize,
}

impl Debug for ResultsSection {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut base = f.debug_set();
        for i in 0..=3 {
            if let Some(e) = self.entries.get(i) {
                base.entry(e);
            }
        }
        if self.entries.len() > 3 {
            base.entry(&"...");
        }
        base.finish()
    }
}

#[serde_with::skip_serializing_none]
#[derive(Serialize, Default, Debug, Clone, utoipa::ToSchema)]
struct ResultEntry {
    #[serde(skip)]
    hit: MSHit,
    /// The id of the location
    #[schema(example = "5510.03.002")]
    id: String,
    /// the type of the site/building
    #[schema(example = "room")]
    r#type: String,
    /// Subtext to show below the search result.
    ///
    /// Usually contains the context of where this rooms is located in.
    /// Currently not highlighted.
    #[schema(example = "5510.03.002 (\x19MW\x17 2001, Empore)")]
    name: String,
    /// Subtext to show below the search result.
    ///
    /// Usually contains the context of where this rooms is located in.
    /// Currently not highlighted.
    #[schema(example = "Maschinenwesen (MW)")]
    subtext: String,
    /// Subtext to show below the search (by default in bold and after the non-bold subtext).
    ///
    /// Usually contains the arch-id of the room, which is another common room id format, and supports highlighting.
    #[schema(example = "3002@5510")]
    subtext_bold: Option<String>,
    /// This is an optional feature, that is only supported for some rooms.
    ///
    /// It might be displayed instead or before the name, to show that a different room id format has matched, that was probably used.
    /// See the image below for an example.
    /// It will be cropped to a maximum length to not take too much space in UIs.
    /// Supports highlighting.
    parsed_id: Option<String>,
}

#[tracing::instrument]
pub async fn address_search(q: &str) -> LimitedVec<ResultsSection> {
    let results = match Nominatim::address_search(q).await {
        Ok(r) => r.0,
        Err(e) => {
            error!(error = ?e, "Error searching for addresses");
            return LimitedVec(vec![]);
        }
    };
    let num_results = results.len();
    let section = ResultsSection {
        facet: ResultFacet::Addresses,
        entries: results
            .into_iter()
            .map(|r| {
                let subtext = r.address.serialise();
                ResultEntry {
                    hit: Default::default(),
                    id: format!("osm_{}", r.osm_id),
                    r#type: r.address_type,
                    name: r.address.road.unwrap_or(r.name),
                    subtext,
                    subtext_bold: None,
                    parsed_id: None,
                }
            })
            .collect(),
        n_visible: num_results.min(15),
        estimated_total_hits: num_results,
    };
    LimitedVec::from(vec![section])
}

#[tracing::instrument(skip(client))]
pub async fn do_geoentry_search(
    client: &Client,
    q: &str,
    highlighting: Highlighting,
    limits: Limits,
) -> LimitedVec<ResultsSection> {
    let parsed_input = ParsedQuery::from(q);

    let query = parsed_input
        .tokens
        .clone()
        .into_iter()
        .map(|s| match s {
            TextToken::Text(t) => t,
            TextToken::SplittableText((t1, t2)) => format!("{t1} {t2} {t1}{t2}"),
        })
        .collect::<Vec<String>>()
        .join(" ");
    let mut query = GeoEntryQuery::from((client, query, &limits, &highlighting));
    for sort in parsed_input.sorting.as_meilisearch_sorting() {
        query.with_sorting(sort);
    }
    if !parsed_input.filters.is_empty() {
        query.with_filtering(parsed_input.filters.as_meilisearch_filters());
    }

    let Ok(response) = query.execute().await else {
        // error should be serde_json::error
        error!("Error searching for results");
        return LimitedVec(vec![]);
    };
    let (section_buildings, mut section_rooms) = merger::merge_search_results(
        &limits,
        response.results.first().unwrap(),
        response.results.get(1).unwrap(),
        response.results.get(2).unwrap(),
    );
    let visitor = formatter::RoomVisitor::from((parsed_input, highlighting));
    section_rooms
        .entries
        .iter_mut()
        .for_each(|r| visitor.visit(r));

    match section_buildings.n_visible {
        0 => LimitedVec(vec![section_rooms, section_buildings]),
        _ => LimitedVec(vec![section_buildings, section_rooms]),
    }
}

#[cfg(test)]
mod test {
    use std::fmt::{Display, Formatter};

    use super::*;
    use crate::setup::tests::MeiliSearchTestContainer;

    #[derive(serde::Deserialize)]
    struct TestQuery {
        target: String,
        query: String,
        among: Option<usize>,
        comment: Option<String>,
    }

    impl TestQuery {
        fn load_good() -> Vec<Self> {
            serde_yaml::from_str(include_str!("test-queries.good.yaml")).unwrap()
        }
        fn load_bad() -> Vec<Self> {
            serde_yaml::from_str(include_str!("test-queries.bad.yaml")).unwrap()
        }
        fn actual_matches_among(&self, actual: &[ResultsSection]) -> bool {
            let among = self.among.unwrap_or(1);
            let mut acceptable_range = actual.iter().flat_map(|r| r.entries.clone()).take(among);
            acceptable_range.any(|r| r.id == self.target)
        }
        async fn search(&self, client: &Client) -> Vec<ResultsSection> {
            do_geoentry_search(
                client,
                &self.query,
                Highlighting::default(),
                Limits::default(),
            )
            .await
            .0
        }
    }
    impl Display for TestQuery {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            write!(
                f,
                "'{query}' should get '{target}' in {among} results",
                query = self.query,
                target = self.target,
                among = self.among.unwrap_or(1),
            )?;
            if let Some(comment) = &self.comment {
                write!(f, " # {comment}")?;
            }
            Ok(())
        }
    }
    #[tokio::test]
    #[tracing_test::traced_test]
    async fn test_good_queries() {
        let ms = MeiliSearchTestContainer::new().await;
        crate::setup::meilisearch::load_data(&ms.client)
            .await
            .unwrap();
        for query in TestQuery::load_good() {
            let actual = query.search(&ms.client).await;
            assert!(
                query.actual_matches_among(&actual),
                "{query}\n\
                Since it can't, please move it to .bad list"
            );

            // redacting estimatedTotalHits is because this can change quite easily and without a good reason
            insta::with_settings!({
                info => &format!("{query}"),
                description => query.comment.unwrap_or_default(),
            }, {
                insta::assert_yaml_snapshot!(actual, { ".**.estimatedTotalHits" => "[estimatedTotalHits]"});
            });
        }
    }

    #[tokio::test]
    #[tracing_test::traced_test]
    async fn test_bad_queries() {
        let ms = MeiliSearchTestContainer::new().await;
        crate::setup::meilisearch::load_data(&ms.client)
            .await
            .unwrap();
        for query in TestQuery::load_bad() {
            let actual = query.search(&ms.client).await;
            assert!(
                !query.actual_matches_among(&actual),
                "{query}\n\
                Since it can, please move it to .good list"
            );

            // redacting estimatedTotalHits is because this can change quite easily and without a good reason
            insta::with_settings!({
                info => &format!("{query}"),
                description => query.comment.unwrap_or_default(),
            }, {
                insta::assert_yaml_snapshot!(actual, { ".**.estimatedTotalHits" => "[estimatedTotalHits]"});
            });
        }
    }
}
