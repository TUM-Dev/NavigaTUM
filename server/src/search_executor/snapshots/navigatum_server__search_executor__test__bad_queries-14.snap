---
source: src/search_executor/mod.rs
description: "there are two basic lab course rooms, this is one of them"
expression: actual
info: "'praktikumsraum mi' should get '5604.EG.038' in 2 results # there are two basic lab course rooms, this is one of them"
---
- facet: rooms
  entries:
    - id: 5605.01.012
      type: room
      name: "5605.01.012 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 01.05.012@5605
    - id: 5605.01.013
      type: room
      name: "5605.01.013 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 01.05.013@5605
    - id: 5605.02.014
      type: room
      name: "5605.02.014 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 02.05.014@5605
    - id: 5605.02.033
      type: room
      name: "5605.02.033 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 02.05.033@5605
    - id: 5605.03.012
      type: room
      name: "5605.03.012 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 03.05.012@5605
    - id: 5605.03.014
      type: room
      name: "5605.03.014 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 03.05.014@5605
    - id: 5605.03.057
      type: room
      name: "5605.03.057 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 03.05.057@5605
    - id: 5607.01.012
      type: room
      name: "5607.01.012 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 01.07.012@5607
    - id: 5608.01.011
      type: room
      name: "5608.01.011 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 01.08.011@5608
    - id: 5608.01.020
      type: room
      name: "5608.01.020 (\u0019Praktikumsraum\u0017)"
      subtext: "garching, Mathe/Info (MI)"
      subtext_bold: 01.08.020@5608
  n_visible: 10
  estimatedTotalHits: "[estimatedTotalHits]"
- facet: sites_buildings
  entries: []
  n_visible: 0
  estimatedTotalHits: "[estimatedTotalHits]"
